lib/neovim-plugin.nix

```diff
@@ -91,18 +101,27 @@ with lib;
       config =
         let
           cfg = config.${namespace}.${name};
-          extraConfigNamespace = if isColorscheme then "extraConfigLuaPre" else "extraConfigLua";
         in
         mkIf cfg.enable (mkMerge [
           {
             extraPlugins = (optional installPackage cfg.package) ++ extraPlugins;
             inherit extraPackages;
           }
-          (optionalAttrs callSetup {
-            ${extraConfigNamespace} = ''
-              require('${luaName}')${setup}(${optionalString (cfg ? settings) (helpers.toLuaObject cfg.settings)})
-            '';
-          })
+          (
+            let
+              setupCode = ''
+                require('${luaName}')${setup}(${optionalString (cfg ? settings) (helpers.toLuaObject cfg.settings)})
+              '';
+            in
+            if !skipConfigAttrs then
+              (optionalAttrs callSetup { ${namespace}.${name}.config.init = setupCode; })
+              // (optionalAttrs (configLocation != null) (setAttrByPath (toList configLocation) cfg.config.final))
+            else
+              assert assertMsg (
+                callSetup -> configLocation != null
+              ) "When a plugin has no config attrs and has a setup function it must have a config location";
+              optionalAttrs callSetup (setAttrByPath (toList configLocation) setupCode)
+          )
```

Comment:
Something about this feels a little messy. Can't quite put my finger on it though... Maybe a combination of minor things?

- why use `!` in an if-then-else when we can just flip it?
- The `optionalAttrs // optionalAttrs` could fail if `configLocation` ever starts with `namespace`
  - We could use a recursive update, but probably better to just use `mkMerge`
- Having nested conditional branches _within_ a `mkMerge` feels messy, probably better to just have all variations listed directly in the `mkMerge`?

Perhaps something like:
```nix
config =
  let
    setupLine = ''
      require('${luaName}')${setup}(${optionalString (cfg ? settings) (helpers.toLuaObject cfg.settings)})
    '';
  in
  assert assertMsg (
    # FIXME: check this logic
    (skipConfigAttrs && callSetup) -> configLocation != null
  ) "When a plugin has no config attrs and has a setup function it must have a config location";
  mkMerge [
    (optionalAttrs (callSetup && !skipConfigAttrs) { ${namespace}.${name}.config.init = setupCode; })
    (optionalAttrs (configLocation != null && !skipConfigAttrs) (setAttrByPath (toList configLocation) cfg.config.final))
    # etc
  ];
```
?

Reply:
I wanted to factor the skipConfigAttrs calls, because it became quite complicated to follow the flow with everthing in the same condition, as you have a `callSetup && skipConfigAttrs`, a `callSetup && !skipConfigAttrs` and a `configLocation != null && !skipConfigAttrs`

After 24.11 we should be able to remove the skipConfigAttrs, and it would simplify a lot the module

Reply:
Hm, I wonder if another way to approach this would be to concat `optionals` lists together, potentially containing optionalAttrs?

```nix
mkMerge (
  optionals skipConfigAttrs [
    (optionalAttrs callSetup (setAttrByPath (toList configLocation) setupCode)
  ]
  ++ optionals (!skipConfigAttrs) [
    (optionalAttrs callSetup { ${namespace}.${name}.config.init = setupCode; })
    (optionalAttrs (configLocation != null) {})
  ]
);
```

But I think you're right, half of the issue is just that it is a messy problem to solve, so I shouldn't be surprised that it ends up looking messy :joy: